sdk/identity/identity/src/credentials/onBehalfOfCredential.ts

```diff
@@ -0,0 +1,144 @@
+   * const client = new KeyClient("vault-url", tokenCredential);
+   *
+   * await client.getKey("key-name", { authenticationOptions: { userAssertion } });
+   * ```
+   *
+   * @param tenantId - The Azure Active Directory tenant (directory) ID.
+   * @param clientId - The client (application) ID of an App Registration in the tenant.
+   * @param clientSecret - A client secret that was generated for the App Registration.
+   * @param userAssertionToken - The user assertion to be used.
+   * @param options - Options for configuring the client which makes the authentication request.
+   */
+  constructor(
+    tenantId: string,
+    clientId: string,
+    clientSecret: string,
+    userAssertionToken: string,
+    options?: OnBehalfOfCredentialOptions
+  );
+  /**
+   * Creates an instance of the {@link OnBehalfOfCredential} with the details
+   * needed to authenticate against Azure Active Directory with a client certificate, and an user assertion.
+   *
+   * Example using the `KeyClient` from [\@azure/keyvault-keys](https://www.npmjs.com/package/\@azure/keyvault-keys):
+   *
+   * ```ts
+   * const certificatePath = "path/to/certificate.pem";
+   * const tokenCredential = new OnBehalfOfCredential(tenantId, clientId, certificatePath, "access-token");
+   * const client = new KeyClient("vault-url", tokenCredential);
+   *
+   * await client.getKey("key-name", { authenticationOptions: { userAssertion } });
+   * ```
+   *
+   * @param tenantId - The Azure Active Directory tenant (directory) ID.
+   * @param clientId - The client (application) ID of an App Registration in the tenant.
+   * @param certificatePath - The path to a PEM-encoded public/private key certificate on the filesystem.
+   * @param userAssertionToken - The user assertion to be used.
+   * @param options - Options for configuring the client which makes the authentication request.
+   */
+  constructor(
+    tenantId: string,
+    clientId: string,
+    certificatePath: string,
+    userAssertionToken: string,
+    options?: OnBehalfOfCredentialOptions
+  );
+
+  /**
+   * Creates an instance of the {@link OnBehalfOfCredential} with the details
+   * needed to authenticate against Azure Active Directory with a client
+   * secret (or certificate), and an user assertion.
+   *
+   * @param tenantId - The Azure Active Directory tenant (directory) ID.
+   * @param clientId - The client (application) ID of an App Registration in the tenant.
+   * @param clientSecretOrCertificatePath - A client secret, or a path to a certificate, that was generated for the App Registration.
+   * @param options - Options for configuring the client which makes the authentication request.
+   */
+  constructor(
+    private tenantId: string,
+    private clientId: string,
+    private clientSecretOrCertificatePath: string,
+    private userAssertionToken: string,
+    private options: OnBehalfOfCredentialOptions = {}
```

Comment:
what if we fixed the ambiguous argument problem by taking an input object instead?

```ts
interface OnBehalfOfCredentialConfiguration {
    tenantId: string;
    clientId: string;
    clientSecret?: string;
    certificatePath?: string;
    userAssertionToken: string;
}

// or alternatively we could get fancy and make the fact one is required explicit:

type OnBehalfOfCredentialConfiguration  = OnBehalfOfCredentialSecretConfiguration  | OnBehalfOfCredentialCertificateConfiguration;

interface OnBehalfOfCredentialSecretConfiguration {
    tenantId: string;
    clientId: string;
    clientSecret: string;
    userAssertionToken: string;
}

interface OnBehalfOfCredentialCertificateConfiguration {
    tenantId: string;
    clientId: string;
    certificatePath: string;
    userAssertionToken: string;
}

```


```suggestion
  constructor(
    private configuration: OnBehalfOfCredentialConfiguration,
    private options: OnBehalfOfCredentialOptions = {}
```

Reply:
Can do!